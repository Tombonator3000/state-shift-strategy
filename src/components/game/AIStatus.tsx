import { Card } from '@/components/ui/card';
import { Badge } from '@/components/ui/badge';
import { Progress } from '@/components/ui/progress';
import { Bot, Brain, Zap, Shield, Target } from 'lucide-react';
import { type AIDifficulty } from '@/data/aiStrategy';

interface AIStatusProps {
  difficulty: AIDifficulty;
  personalityName?: string;
  isThinking?: boolean;
  currentPlayer: 'human' | 'ai';
  aiControlledStates: number;
  assessmentText?: string;
}

const AIStatus = ({ 
  difficulty, 
  personalityName, 
  isThinking = false, 
  currentPlayer,
  aiControlledStates,
  assessmentText 
}: AIStatusProps) => {
  const getDifficultyColor = (diff: AIDifficulty) => {
    switch (diff) {
      case 'easy': return 'bg-green-900/50 text-green-400 border-green-600';
      case 'medium': return 'bg-yellow-900/50 text-yellow-400 border-yellow-600';
      case 'hard': return 'bg-red-900/50 text-red-400 border-red-600';
      case 'legendary': return 'bg-purple-900/50 text-purple-400 border-purple-600';
    }
  };

  const getDifficultyIcon = (diff: AIDifficulty) => {
    switch (diff) {
      case 'easy': return <Bot size={16} />;
      case 'medium': return <Brain size={16} />;
      case 'hard': return <Zap size={16} />;
      case 'legendary': return <Target size={16} />;
    }
  };

  return (
    <Card className="p-3 bg-gray-900 border-gray-700">
      <div className="flex items-center justify-between mb-2">
        <div className="flex items-center gap-2">
          {getDifficultyIcon(difficulty)}
          <h3 className="font-bold text-sm text-white font-mono">
            AI OPPONENT
          </h3>
        </div>
        <Badge className={`text-xs font-bold ${getDifficultyColor(difficulty)}`}>
          {difficulty.toUpperCase()}
        </Badge>
      </div>

      {personalityName && (
        <div className="text-xs text-gray-400 mb-2 font-mono">
          {personalityName}
        </div>
      )}

      <div className="space-y-2">
        {/* AI Status Indicator */}
        <div className="flex items-center gap-2">
          <div className={`w-2 h-2 rounded-full ${
            currentPlayer === 'ai' 
              ? 'bg-red-500 animate-pulse' 
              : 'bg-gray-600'
          }`} />
          <span className="text-xs text-gray-300">
            {currentPlayer === 'ai' 
              ? (isThinking ? 'Calculating...' : 'Active Turn') 
              : 'Waiting'
            }
          </span>
        </div>

        {/* Territory Control */}
        <div className="flex items-center justify-between text-xs">
          <span className="text-gray-400">Territory:</span>
          <span className="text-red-400 font-bold">
            {aiControlledStates} states
          </span>
        </div>

        {/* AI Assessment */}
        {assessmentText && (
          <div className="text-xs text-gray-500 italic border-t border-gray-700 pt-2 truncate">
            {assessmentText}
          </div>
        )}

        {/* Thinking Animation */}
        {isThinking && currentPlayer === 'ai' && (
          <div className="flex items-center gap-1 text-xs text-red-400">
            <div className="flex gap-1">
              <div className="w-1 h-1 bg-red-400 rounded-full animate-bounce" style={{ animationDelay: '0ms' }} />
              <div className="w-1 h-1 bg-red-400 rounded-full animate-bounce" style={{ animationDelay: '150ms' }} />
              <div className="w-1 h-1 bg-red-400 rounded-full animate-bounce" style={{ animationDelay: '300ms' }} />
            </div>
            <span>Processing strategy...</span>
          </div>
        )}
      </div>
    </Card>
  );
};

export default AIStatus;