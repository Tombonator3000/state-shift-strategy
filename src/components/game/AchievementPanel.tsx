import { useState, useEffect } from 'react';
import { Card } from '@/components/ui/card';
import { Button } from '@/components/ui/button';
import { Badge } from '@/components/ui/badge';
import { Progress } from '@/components/ui/progress';
import { Tabs, TabsContent, TabsList, TabsTrigger } from '@/components/ui/tabs';
import { ScrollArea } from '@/components/ui/scroll-area';
import { X, Trophy, Star, Target, Zap, Users, BookOpen, Eye, Download, Upload, RotateCcw } from 'lucide-react';
import { ACHIEVEMENTS, type Achievement } from '@/data/achievementSystem';
import { useAchievements } from '@/contexts/AchievementContext';
import { useToast } from '@/hooks/use-toast';

interface AchievementsSectionProps {
  onClose?: () => void;
  className?: string;
  showCloseButton?: boolean;
}

export const AchievementsSection = ({
  onClose,
  className,
  showCloseButton = false,
}: AchievementsSectionProps) => {
  const [selectedAchievement, setSelectedAchievement] = useState<Achievement | null>(null);
  const [filter, setFilter] = useState<string>('all');

  const {
    manager,
    stats,
    unlockedAchievements,
    lockedAchievements,
    newlyUnlocked,
    exportData,
    importData,
    resetProgress,
    clearNewlyUnlocked
  } = useAchievements();

  const { toast } = useToast();

  const progress = manager.getProgress();

  useEffect(() => {
    if (newlyUnlocked.length > 0) {
      clearNewlyUnlocked();
    }
  }, [newlyUnlocked, clearNewlyUnlocked]);

  const getCategoryIcon = (category: string) => {
    switch (category) {
      case 'victory': return <Trophy size={16} />;
      case 'mastery': return <Star size={16} />;
      case 'discovery': return <Eye size={16} />;
      case 'challenge': return <Target size={16} />;
      case 'social': return <Users size={16} />;
      case 'collection': return <BookOpen size={16} />;
      default: return <Zap size={16} />;
    }
  };

  const getRarityColor = (rarity: string) => {
    switch (rarity) {
      case 'common': return 'text-gray-400 bg-gray-900/20 border-gray-600';
      case 'uncommon': return 'text-green-400 bg-green-900/20 border-green-600';
      case 'rare': return 'text-blue-400 bg-blue-900/20 border-blue-600';
      case 'legendary': return 'text-purple-400 bg-purple-900/20 border-purple-600';
      default: return 'text-gray-400 bg-gray-900/20 border-gray-600';
    }
  };

  const filteredUnlocked = filter === 'all'
    ? unlockedAchievements
    : unlockedAchievements.filter(a => a.category === filter);

  const filteredLocked = filter === 'all'
    ? lockedAchievements
    : lockedAchievements.filter(a => a.category === filter);

  const exportProgress = () => {
    const data = exportData();
    const blob = new Blob([JSON.stringify(data, null, 2)], { type: 'application/json' });
    const url = URL.createObjectURL(blob);
    const link = document.createElement('a');
    link.href = url;
    link.download = `shadow-government-progress-${new Date().toISOString().split('T')[0]}.json`;
    link.click();
    URL.revokeObjectURL(url);

    toast({
      title: 'Progress Exported',
      description: 'Your achievements and statistics have been downloaded',
    });
  };

  const importProgress = () => {
    const input = document.createElement('input');
    input.type = 'file';
    input.accept = '.json';
    input.onchange = (event: any) => {
      const file = event.target.files[0];
      if (file) {
        const reader = new FileReader();
        reader.onload = (e) => {
          try {
            const data = JSON.parse(e.target?.result as string);
            importData(data);
          } catch (error) {
            toast({
              title: 'Import Failed',
              description: 'Invalid file format',
              variant: 'destructive',
            });
          }
        };
        reader.readAsText(file);
      }
    };
    input.click();
  };

  const handleResetProgress = () => {
    if (window.confirm('Are you sure you want to reset all achievements and statistics? This cannot be undone.')) {
      resetProgress();
    }
  };

  return (
    <div className={`flex h-full flex-col ${className ?? ''}`}>
      <div className="flex items-center justify-between border-b border-gray-700 p-4">
        <div className="flex items-center gap-3">
          <Trophy size={24} className="text-yellow-400" />
          <div>
            <h2 className="text-xl font-bold text-white font-mono">ACHIEVEMENTS</h2>
            <div className="text-sm text-gray-400">
              {progress.unlocked}/{progress.total} unlocked • {progress.totalPoints} points • {progress.rank}
            </div>
          </div>
        </div>
        <div className="flex items-center gap-2">
          <Button
            onClick={exportProgress}
            variant="outline"
            size="sm"
            className="text-green-400 border-green-600 hover:bg-green-900/20"
          >
            <Download size={16} className="mr-1" />
            Export
          </Button>
          <Button
            onClick={importProgress}
            variant="outline"
            size="sm"
            className="text-blue-400 border-blue-600 hover:bg-blue-900/20"
          >
            <Upload size={16} className="mr-1" />
            Import
          </Button>
          <Button
            onClick={handleResetProgress}
            variant="outline"
            size="sm"
            className="text-red-400 border-red-600 hover:bg-red-900/20"
          >
            <RotateCcw size={16} className="mr-1" />
            Reset
          </Button>
          {showCloseButton && onClose && (
            <Button
              onClick={onClose}
              variant="outline"
              size="sm"
              className="text-gray-400 border-gray-600"
            >
              <X size={16} />
            </Button>
          )}
        </div>
      </div>

      <div className="flex-1 overflow-hidden p-4">
        <Tabs defaultValue="achievements" className="h-full">
          <TabsList className="mb-4 grid w-full grid-cols-3 bg-gray-800">
            <TabsTrigger value="achievements">Achievements</TabsTrigger>
            <TabsTrigger value="statistics">Statistics</TabsTrigger>
            <TabsTrigger value="progress">Progress</TabsTrigger>
          </TabsList>

          <TabsContent value="achievements" className="h-full">
            <div className="grid h-full grid-cols-1 gap-4 lg:grid-cols-3">
              <div className="lg:col-span-2">
                <Card className="h-full border-gray-700 bg-gray-800 p-4">
                  <div className="mb-4 flex items-center justify-between">
                    <h3 className="text-lg font-semibold text-white">Achievement List</h3>
                    <select
                      value={filter}
                      onChange={(e) => setFilter(e.target.value)}
                      className="rounded border border-gray-600 bg-gray-700 px-3 py-1 text-sm text-white"
                    >
                      <option value="all">All Categories</option>
                      <option value="victory">Victory</option>
                      <option value="mastery">Mastery</option>
                      <option value="discovery">Discovery</option>
                      <option value="challenge">Challenge</option>
                      <option value="social">Social</option>
                      <option value="collection">Collection</option>
                    </select>
                  </div>

                  <ScrollArea className="h-[60vh]">
                    <div className="space-y-4">
                      {filteredUnlocked.length > 0 && (
                        <div>
                          <h4 className="mb-3 text-sm font-bold uppercase tracking-wide text-green-400">
                            Unlocked ({filteredUnlocked.length})
                          </h4>
                          <div className="space-y-2">
                            {filteredUnlocked.map(achievement => (
                              <div
                                key={achievement.id}
                                className={`rounded p-3 transition-colors cursor-pointer ${
                                  selectedAchievement?.id === achievement.id
                                    ? 'bg-blue-900/30 border border-blue-600'
                                    : 'bg-gray-700 hover:bg-gray-600'
                                }`}
                                onClick={() => setSelectedAchievement(achievement)}
                              >
                                <div className="flex items-center gap-3">
                                  <div className="text-2xl">{achievement.icon}</div>
                                  <div className="flex-1">
                                    <div className="mb-1 flex items-center gap-2">
                                      <h5 className="font-semibold text-white">{achievement.name}</h5>
                                      <Badge className={getRarityColor(achievement.rarity)}>
                                        {achievement.rarity}
                                      </Badge>
                                      <Badge className="bg-yellow-900/20 text-yellow-400">
                                        {achievement.points}pts
                                      </Badge>
                                    </div>
                                    <p className="text-sm text-gray-300">{achievement.description}</p>
                                    <div className="mt-1 flex items-center gap-2">
                                      {getCategoryIcon(achievement.category)}
                                      <span className="text-xs capitalize text-gray-500">{achievement.category}</span>
                                      <span className="text-xs text-green-400">✓ Completed</span>
                                    </div>
                                  </div>
                                </div>
                              </div>
                            ))}
                          </div>
                        </div>
                      )}

                      {filteredLocked.length > 0 && (
                        <div>
                          <h4 className="mb-3 text-sm font-bold uppercase tracking-wide text-gray-400">
                            Locked ({filteredLocked.length})
                          </h4>
                          <div className="space-y-2">
                            {filteredLocked.map(achievement => (
                              <div
                                key={achievement.id}
                                className={`rounded p-3 transition-colors opacity-60 cursor-pointer ${
                                  selectedAchievement?.id === achievement.id
                                    ? 'bg-blue-900/30 border border-blue-600'
                                    : 'bg-gray-700 hover:bg-gray-600'
                                }`}
                                onClick={() => setSelectedAchievement(achievement)}
                              >
                                <div className="flex items-center gap-3">
                                  <div className="text-2xl grayscale">{achievement.icon}</div>
                                  <div className="flex-1">
                                    <div className="mb-1 flex items-center gap-2">
                                      <h5 className="font-semibold text-gray-300">{achievement.name}</h5>
                                      <Badge className={getRarityColor(achievement.rarity)}>
                                        {achievement.rarity}
                                      </Badge>
                                      <Badge className="bg-yellow-900/20 text-yellow-400">
                                        {achievement.points}pts
                                      </Badge>
                                    </div>
                                    <p className="text-sm text-gray-400">{achievement.description}</p>
                                    <div className="mt-1 flex items-center gap-2">
                                      {getCategoryIcon(achievement.category)}
                                      <span className="text-xs capitalize text-gray-500">{achievement.category}</span>
                                      <span className="text-xs text-gray-500">🔒 Locked</span>
                                    </div>
                                  </div>
                                </div>
                              </div>
                            ))}
                          </div>
                        </div>
                      )}
                    </div>
                  </ScrollArea>
                </Card>
              </div>

              <div>
                <Card className="h-full border-gray-700 bg-gray-800 p-4">
                  <h3 className="mb-4 text-lg font-semibold text-white">
                    {selectedAchievement ? 'Achievement Details' : 'Select Achievement'}
                  </h3>
                  {selectedAchievement ? (
                    <ScrollArea className="h-[60vh]">
                      <div className="space-y-4">
                        <div className="text-center">
                          <div className="mb-2 text-6xl">{selectedAchievement.icon}</div>
                          <h4 className="text-xl font-bold text-white">{selectedAchievement.name}</h4>
                          <p className="mt-1 text-sm text-gray-300">{selectedAchievement.description}</p>
                        </div>

                        <div className="space-y-3">
                          <div className="flex items-center justify-between">
                            <span className="text-sm text-gray-400">Category:</span>
                            <div className="flex items-center gap-1">
                              {getCategoryIcon(selectedAchievement.category)}
                              <span className="text-sm capitalize text-white">{selectedAchievement.category}</span>
                            </div>
                          </div>

                          <div className="flex items-center justify-between">
                            <span className="text-sm text-gray-400">Rarity:</span>
                            <Badge className={getRarityColor(selectedAchievement.rarity)}>
                              {selectedAchievement.rarity}
                            </Badge>
                          </div>

                          <div className="flex items-center justify-between">
                            <span className="text-sm text-gray-400">Points:</span>
                            <span className="text-sm font-bold text-yellow-400">{selectedAchievement.points}</span>
                          </div>

                          <div className="flex items-center justify-between">
                            <span className="text-sm text-gray-400">Status:</span>
                            <Badge className={
                              unlockedAchievements.find(a => a.id === selectedAchievement.id)
                                ? 'text-green-400 bg-green-900/20'
                                : 'text-red-400 bg-red-900/20'
                            }>
                              {unlockedAchievements.find(a => a.id === selectedAchievement.id) ? '✓ Unlocked' : '🔒 Locked'}
                            </Badge>
                          </div>

                          {selectedAchievement.requirements && (
                            <div>
                              <div className="mb-2 text-sm text-gray-400">Requirements:</div>
                              <div className="space-y-1">
                                {selectedAchievement.requirements.conditions.map((condition, index) => (
                                  <div key={index} className="rounded bg-gray-700 p-2 text-xs text-gray-500">
                                    {condition.key.replace(/_/g, ' ')}: {condition.operator || '=='} {condition.value}
                                  </div>
                                ))}
                              </div>
                            </div>
                          )}

                          {selectedAchievement.rewards && (
                            <div>
                              <div className="mb-2 text-sm text-gray-400">Rewards:</div>
                              <div className="space-y-1">
                                {selectedAchievement.rewards.title && (
                                  <div className="text-xs text-yellow-400">Title: {selectedAchievement.rewards.title}</div>
                                )}
                                {selectedAchievement.rewards.unlockTutorial && (
                                  <div className="text-xs text-blue-400">
                                    Unlocks: {selectedAchievement.rewards.unlockTutorial} tutorial
                                  </div>
                                )}
                              </div>
                            </div>
                          )}

                          {selectedAchievement.hidden && (
                            <div className="rounded border border-purple-600 bg-purple-900/20 p-3">
                              <div className="text-sm font-bold text-purple-400">Hidden Achievement</div>
                              <div className="mt-1 text-xs text-purple-300">
                                This is a secret achievement that was discovered through special actions.
                              </div>
                            </div>
                          )}
                        </div>
                      </div>
                    </ScrollArea>
                  ) : (
                    <div className="py-8 text-center text-gray-500">
                      <Trophy size={48} className="mx-auto mb-4 opacity-50" />
                      <div className="mb-2 text-lg font-medium">Achievement Details</div>
                      <div className="text-sm">
                        Click on an achievement to view detailed information
                      </div>
                    </div>
                  )}
                </Card>
              </div>
            </div>
          </TabsContent>

          <TabsContent value="statistics" className="mt-4">
            <div className="grid grid-cols-1 gap-4 md:grid-cols-2 lg:grid-cols-4">
              <Card className="border-gray-700 bg-gray-800 p-4">
                <div className="text-2xl font-bold text-white">{stats.total_games_played}</div>
                <div className="text-sm text-gray-400">Games Played</div>
              </Card>
              <Card className="border-gray-700 bg-gray-800 p-4">
                <div className="text-2xl font-bold text-green-400">{stats.games_won}</div>
                <div className="text-sm text-gray-400">Games Won</div>
              </Card>
              <Card className="border-gray-700 bg-gray-800 p-4">
                <div className="text-2xl font-bold text-yellow-400">{stats.max_win_streak}</div>
                <div className="text-sm text-gray-400">Best Win Streak</div>
              </Card>
              <Card className="border-gray-700 bg-gray-800 p-4">
                <div className="text-2xl font-bold text-blue-400">{progress.totalPoints}</div>
                <div className="text-sm text-gray-400">Achievement Points</div>
              </Card>
              <Card className="border-gray-700 bg-gray-800 p-4">
                <div className="text-2xl font-bold text-purple-400">{stats.legendary_cards_played}</div>
                <div className="text-sm text-gray-400">Legendary Cards</div>
              </Card>
              <Card className="border-gray-700 bg-gray-800 p-4">
                <div className="text-2xl font-bold text-red-400">{stats.max_states_controlled_single_game}</div>
                <div className="text-sm text-gray-400">Max States Controlled</div>
              </Card>
              <Card className="border-gray-700 bg-gray-800 p-4">
                <div className="text-2xl font-bold text-cyan-400">
                  {stats.fastest_victory_turns === 999 ? '—' : stats.fastest_victory_turns}
                </div>
                <div className="text-sm text-gray-400">Fastest Victory (turns)</div>
              </Card>
              <Card className="border-gray-700 bg-gray-800 p-4">
                <div className="text-2xl font-bold text-orange-400">{Math.round(stats.total_play_time_minutes / 60)}h</div>
                <div className="text-sm text-gray-400">Play Time</div>
              </Card>
              <Card className="border-gray-700 bg-gray-800 p-4 md:col-span-2">
                <div className="grid grid-cols-1 gap-4 sm:grid-cols-3">
                  <div>
                    <div className="text-2xl font-bold text-fuchsia-400">{stats.total_combos_executed}</div>
                    <div className="text-sm text-gray-400">Combos Triggered</div>
                  </div>
                  <div>
                    <div className="text-2xl font-bold text-sky-400">{stats.highest_combo_chain}</div>
                    <div className="text-sm text-gray-400">Best Combo Chain</div>
                  </div>
                  <div>
                    <div className="text-2xl font-bold text-amber-400">{stats.max_combos_in_single_game}</div>
                    <div className="text-sm text-gray-400">Most Combos in a Game</div>
                  </div>
                </div>
              </Card>
              <Card className="border-gray-700 bg-gray-800 p-4 md:col-span-2">
                <h4 className="mb-3 text-sm font-semibold text-gray-200">Combo Category Breakdown</h4>
                <div className="grid grid-cols-1 gap-2 text-sm text-gray-300 sm:grid-cols-2">
                  <div className="flex items-center justify-between">
                    <span>Sequence</span>
                    <span className="font-semibold text-white">{stats.total_sequence_combos}</span>
                  </div>
                  <div className="flex items-center justify-between">
                    <span>Count</span>
                    <span className="font-semibold text-white">{stats.total_count_combos}</span>
                  </div>
                  <div className="flex items-center justify-between">
                    <span>Threshold</span>
                    <span className="font-semibold text-white">{stats.total_threshold_combos}</span>
                  </div>
                  <div className="flex items-center justify-between">
                    <span>State Control</span>
                    <span className="font-semibold text-white">{stats.total_state_combos}</span>
                  </div>
                  <div className="flex items-center justify-between">
                    <span>Hybrid</span>
                    <span className="font-semibold text-white">{stats.total_hybrid_combos}</span>
                  </div>
                </div>
              </Card>
            </div>
          </TabsContent>

          <TabsContent value="progress" className="mt-4">
            <div className="grid grid-cols-1 gap-4 md:grid-cols-2">
              <Card className="border-gray-700 bg-gray-800 p-4">
                <h3 className="mb-4 text-lg font-semibold text-white">Overall Progress</h3>
                <div className="space-y-4">
                  <div>
                    <div className="mb-1 flex justify-between text-sm">
                      <span>Achievements</span>
                      <span>{progress.unlocked}/{progress.total}</span>
                    </div>
                    <Progress value={progress.completionRate} className="h-3" />
                  </div>
                  <div className="text-center">
                    <div className="text-2xl font-bold text-white">{progress.rank}</div>
                    <div className="text-sm text-gray-400">Current Rank</div>
                  </div>
                </div>
              </Card>

              <Card className="border-gray-700 bg-gray-800 p-4">
                <h3 className="mb-4 text-lg font-semibold text-white">Category Progress</h3>
                <div className="space-y-3">
                  {['victory', 'mastery', 'discovery', 'challenge', 'social', 'collection'].map(category => {
                    const total = ACHIEVEMENTS.filter(a => a.category === category && !a.hidden).length;
                    const unlocked = unlockedAchievements.filter(a => a.category === category).length;
                    const percentage = total > 0 ? Math.round((unlocked / total) * 100) : 0;

                    return (
                      <div key={category}>
                        <div className="mb-1 flex justify-between text-sm">
                          <div className="flex items-center gap-1">
                            {getCategoryIcon(category)}
                            <span className="capitalize">{category}</span>
                          </div>
                          <span>{unlocked}/{total}</span>
                        </div>
                        <Progress value={percentage} className="h-2" />
                      </div>
                    );
                  })}
                </div>
              </Card>
            </div>
          </TabsContent>
        </Tabs>
      </div>
    </div>
  );
};

interface AchievementPanelProps {
  onClose: () => void;
}

const AchievementPanel = ({ onClose }: AchievementPanelProps) => {
  return (
    <div className="fixed inset-0 z-50 flex items-center justify-center bg-black/80 p-4">
      <Card className="h-[90vh] w-full max-w-7xl overflow-hidden border-gray-700 bg-gray-900">
        <AchievementsSection onClose={onClose} showCloseButton className="h-full" />
      </Card>
    </div>
  );
};

export default AchievementPanel;