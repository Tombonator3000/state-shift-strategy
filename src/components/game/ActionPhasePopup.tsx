import React, { useState, useEffect } from 'react';
import { Card } from '@/components/ui/card';

interface ActionPhasePopupProps {
  isVisible: boolean;
  truthLevel: number;
  onClose: () => void;
}

const ActionPhasePopup = ({ isVisible, truthLevel, onClose }: ActionPhasePopupProps) => {
  const [teaser, setTeaser] = useState('');

  useEffect(() => {
    if (isVisible) {
      setTeaser(getRandomTeaser());
    }
  }, [isVisible]);

  useEffect(() => {
    const handleKeyDown = (e: KeyboardEvent) => {
      if (e.key === 'Escape' && isVisible) {
        onClose();
      }
    };

    if (isVisible) {
      document.addEventListener('keydown', handleKeyDown);
      return () => document.removeEventListener('keydown', handleKeyDown);
    }
  }, [isVisible, onClose]);

  const getConspiracyLevel = () => {
    if (truthLevel >= 70) return 'MAXIMUM ALERT';
    if (truthLevel >= 40) return 'ELEVATED PANIC';
    return 'QUIET DESPERATION';
  };

  const getRandomTeaser = () => {
    const teasers = [
      "Moon beams nominal. Hat tinfoil at the ready.",
      "Bat Boy denies involvement â€” which only confirms it.",
      "Emergency horoscopes suggest 'spend IP irresponsibly'.",
      "Local grandmother's shortwave radio picks up alien shopping list.",
      "Three-headed chicken found voting in Ohio â€” twice.",
      "Government weather machine malfunctions: Cats raining dogs.",
      "Mysterious van spotted selling 'truth serum' energy drinks.",
      "Breaking: Pigeons confirmed as government drones, demand workers comp."
    ];
    return teasers[Math.floor(Math.random() * teasers.length)];
  };

  if (!isVisible) return null;

  return (
    <div 
      className="fixed inset-0 bg-black/80 flex items-center justify-center z-50 animate-fade-in"
      onClick={onClose}
    >
      <Card className="bg-newspaper-bg border-4 border-newspaper-border max-w-2xl mx-4 p-0 transform animate-scale-in shadow-2xl">
        <div className="relative">
          {/* Mini Masthead */}
          <div className="bg-newspaper-header p-4 border-b-4 border-newspaper-border text-center">
            <div className="text-xs font-bold tracking-widest opacity-60 mb-1">EST. 1947 â€¢ EYES ONLY</div>
            <h1 className="text-2xl font-black tracking-tight font-serif text-newspaper-text">
              THE WEEKLY PARANOID NEWS
            </h1>
            <div className="text-xs opacity-80 mt-1">SPECIAL BULLETIN</div>
          </div>

          {/* Action Phase Header */}
          <div className="p-6 text-center">
            <div className="flex items-center justify-center gap-2 mb-4">
              <span className="text-4xl font-black text-newspaper-accent font-serif">ACTION PHASE</span>
              <span className="text-3xl text-truth-red animate-pulse">âš¡</span>
            </div>
            <div className="border-b-2 border-newspaper-accent w-32 mx-auto mb-4"></div>
            
            <div className="mb-6">
              <div className="text-sm font-bold text-newspaper-text/80 mb-1">CONSPIRACY LEVEL:</div>
              <div className="text-xl font-black text-truth-red border-2 border-truth-red px-4 py-2 inline-block bg-truth-red/10">
                {getConspiracyLevel()}
              </div>
            </div>

            {/* Rotating Teaser */}
            <div className="bg-newspaper-header/30 p-4 border-l-4 border-newspaper-accent mb-6">
              <div className="text-sm font-bold text-newspaper-text/70 mb-2">ðŸ“° BREAKING WEIRD:</div>
              <div className="text-base italic text-newspaper-text leading-relaxed">
                {teaser}
              </div>
            </div>

            {/* Instructions */}
            <div className="text-sm text-newspaper-text/60 border-t border-newspaper-border pt-4">
              <div className="font-mono">Click anywhere or press ESC to continue...</div>
            </div>
          </div>
        </div>
      </Card>
    </div>
  );
};

export default ActionPhasePopup;