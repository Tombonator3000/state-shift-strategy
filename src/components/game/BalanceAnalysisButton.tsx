import React, { useState } from 'react';
import { Button } from '@/components/ui/button';
import { BarChart3, Calculator } from 'lucide-react';
import FactionBalanceDashboard from './FactionBalanceDashboard';

const BalanceAnalysisButton: React.FC = () => {
  const [showDashboard, setShowDashboard] = useState(false);

  return (
    <>
      <Button
        variant="outline"
        size="sm"
        onClick={() => setShowDashboard(true)}
        className="flex items-center gap-2"
        title="Analyze card balance and faction alignment"
      >
        <BarChart3 size={16} />
        <Calculator size={16} />
        Balance Analysis
      </Button>

      {showDashboard && (
        <FactionBalanceDashboard onClose={() => setShowDashboard(false)} />
      )}
    </>
  );
};

export default BalanceAnalysisButton;