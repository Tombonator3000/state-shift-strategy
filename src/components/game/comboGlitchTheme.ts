import type { ComboTheme } from '@/data/combos/themes';

const clamp01 = (value: number): number => {
  if (Number.isNaN(value)) {
    return 0;
  }
  if (value < 0) {
    return 0;
  }
  if (value > 1) {
    return 1;
  }
  return value;
};

const formatDuration = (value: number): string => {
  if (!Number.isFinite(value)) {
    return '0ms';
  }
  return `${Math.max(0, Math.round(value))}ms`;
};

export interface ApplyComboThemeOptions {
  root: HTMLElement;
  theme: ComboTheme;
  durationMs: number;
  jitterPx: number;
  blurPx: number;
  overlayAlpha: number;
  vignette: boolean;
  halftone: boolean;
  previousThemeId?: string | null;
}

export const applyComboThemeToRoot = ({
  root,
  theme,
  durationMs,
  jitterPx,
  blurPx,
  overlayAlpha,
  vignette,
  halftone,
  previousThemeId,
}: ApplyComboThemeOptions): void => {
  root.classList.add('combo-glitching');

  if (previousThemeId && previousThemeId !== theme.id) {
    root.classList.remove(previousThemeId);
  }

  root.classList.add(theme.id);

  root.style.setProperty('--combo-glitch-duration', formatDuration(durationMs));
  root.style.setProperty('--glitch-jitter', `${Math.max(0, Math.round(jitterPx))}px`);
  root.style.setProperty('--glitch-blur', `${Math.max(0, Math.round(blurPx))}px`);
  root.style.setProperty('--overlay-alpha', clamp01(overlayAlpha).toString());
  root.style.setProperty('--glitch-vignette', vignette ? '1' : '0');
  root.style.setProperty('--glitch-halftone', halftone ? '1' : '0');
};

export const clearComboThemeFromRoot = (root: HTMLElement, currentThemeId?: string | null): void => {
  root.classList.remove('combo-glitching');
  if (currentThemeId) {
    root.classList.remove(currentThemeId);
  }

  root.style.removeProperty('--combo-glitch-duration');
  root.style.removeProperty('--glitch-jitter');
  root.style.removeProperty('--glitch-blur');
  root.style.removeProperty('--overlay-alpha');
  root.style.removeProperty('--glitch-vignette');
  root.style.removeProperty('--glitch-halftone');
};
