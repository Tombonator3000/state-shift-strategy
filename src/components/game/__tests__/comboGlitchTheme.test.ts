import { describe, expect, it } from 'bun:test';
import { applyComboThemeToRoot, clearComboThemeFromRoot } from '@/components/game/comboGlitchTheme';
import { ComboThemeMap } from '@/data/combos/themes';

class MockRootElement {
  private classSet = new Set<string>();
  private styleMap = new Map<string, string>();

  classList = {
    add: (...tokens: string[]) => {
      tokens.forEach(token => this.classSet.add(token));
    },
    remove: (...tokens: string[]) => {
      tokens.forEach(token => this.classSet.delete(token));
    },
  } as unknown as DOMTokenList;

  style = {
    setProperty: (key: string, value: string) => {
      this.styleMap.set(key, value);
    },
    removeProperty: (key: string) => {
      this.styleMap.delete(key);
    },
  } as unknown as CSSStyleDeclaration;

  snapshot() {
    return {
      classes: Array.from(this.classSet).sort(),
      styles: Object.fromEntries(Array.from(this.styleMap.entries()).sort()),
    };
  }
}

describe('comboGlitchTheme helpers', () => {
  it('applies, updates, and clears theme tokens', () => {
    const root = new MockRootElement();

    applyComboThemeToRoot({
      root: root as unknown as HTMLElement,
      theme: ComboThemeMap.MEDIA_WAVE,
      durationMs: 1100,
      jitterPx: 1,
      blurPx: 0,
      overlayAlpha: 0.16,
      vignette: false,
      halftone: true,
      previousThemeId: null,
    });

    expect(root.snapshot()).toMatchSnapshot();

    applyComboThemeToRoot({
      root: root as unknown as HTMLElement,
      theme: ComboThemeMap.COVER_OPERATION,
      durationMs: 1500,
      jitterPx: 2,
      blurPx: 1,
      overlayAlpha: 0.2,
      vignette: true,
      halftone: true,
      previousThemeId: ComboThemeMap.MEDIA_WAVE.id,
    });

    expect(root.snapshot()).toMatchSnapshot();

    clearComboThemeFromRoot(root as unknown as HTMLElement, ComboThemeMap.COVER_OPERATION.id);

    expect(root.snapshot()).toMatchSnapshot();
  });
});
