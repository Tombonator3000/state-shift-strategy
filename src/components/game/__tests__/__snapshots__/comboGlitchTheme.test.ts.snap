// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`comboGlitchTheme helpers applies, updates, and clears theme tokens 1`] = `
{
  "classes": [
    "combo-glitching",
    "interference",
  ],
  "styles": {
    "--combo-glitch-duration": "1100ms",
    "--glitch-blur": "0px",
    "--glitch-halftone": "1",
    "--glitch-jitter": "1px",
    "--glitch-vignette": "0",
    "--overlay-alpha": "0.16",
  },
}
`;

exports[`comboGlitchTheme helpers applies, updates, and clears theme tokens 2`] = `
{
  "classes": [
    "combo-glitching",
    "static",
  ],
  "styles": {
    "--combo-glitch-duration": "1500ms",
    "--glitch-blur": "1px",
    "--glitch-halftone": "1",
    "--glitch-jitter": "2px",
    "--glitch-vignette": "1",
    "--overlay-alpha": "0.2",
  },
}
`;

exports[`comboGlitchTheme helpers applies, updates, and clears theme tokens 3`] = `
{
  "classes": [],
  "styles": {},
}
`;
